src/Features/LanguageServer/Protocol/ILspSolutionProvider.cs

```diff
@@ -4,22 +4,16 @@
 
 using System;
 using System.Collections.Immutable;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer
 {
     internal interface ILspSolutionProvider
     {
         /// <summary>
-        /// Finds the workspace and solution containing the specified document URI
-        /// and returns the documents in that context.
+        /// Finds the document, and the solution it comes from, in any workspace that has registered for LSP
+        /// and including only documents where the client name matches.
         /// </summary>
-        /// <param name="documentUri">the document's file path URI.</param>
-        /// <returns>the documents in the correct workspace and solution context</returns>
-        ImmutableArray<Document> GetDocuments(Uri documentUri);
-
-        /// <summary>
-        /// Return the latest solution from the main workspace that we know about.
-        /// </summary>
-        Solution GetCurrentSolutionForMainWorkspace();
+        (DocumentId?, Solution) FindDocumentAndSolution(TextDocumentIdentifier? textDocument, string? clientName);
```

Comment:
>(DocumentId?, Solution) [](start = 8, length = 23)

I'd return `Document?` instead of the tuple. The caller can get both Id and Solution from Document instance.